p2p/src/network/rpc/p2p_network_rpc_reducer.rs

```diff
@@ -48,53 +98,233 @@ impl P2pNetworkRpcState {
-                    if self.pending.is_none() {
-                        self.pending = Some(header.clone());
+                    if rpc_state.pending.is_none() {
+                        rpc_state.pending = Some(header.clone());
                     } else {
-                        openmina_core::error!(action.time(); "receiving query while another query is pending");
+                        openmina_core::error!(meta.time(); "receiving query while another query is pending");
+                    }
+                }
+
+                rpc_state.incoming.pop_front();
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let network_state: &P2pNetworkState = state.substate()?;
+                let state = network_state
+                    .find_rpc_state(action)
+                    .ok_or_else(|| format!("RPC state not found for action: {:?}", action))?;
+
+                match &message {
+                    RpcMessage::Handshake => {
+                        if !state.is_incoming {
+                            dispatcher.push(P2pChannelsRpcAction::Ready { peer_id: *peer_id });
+                        }
+                    }
+                    RpcMessage::Heartbeat => {}
+                    RpcMessage::Query { header, bytes } => {
+                        if let Err(e) = rpc_query_effects(*peer_id, header, bytes, dispatcher) {
+                            dispatcher.push(P2pDisconnectionAction::Init {
+                                peer_id: *peer_id,
+                                reason: P2pDisconnectionReason::P2pChannelReceiveFailed(
+                                    e.to_string(),
+                                ),
+                            });
+                        }
+                    }
+                    RpcMessage::Response {
+                        header: ResponseHeader { id },
+                        bytes,
+                    } => {
+                        let query_header = match state.pending.as_ref() {
+                            Some(header) if &header.id == id => header.clone(),
+                            Some(header) => {
+                                error!(meta.time(); "received response with it {} while expecting {id}", header.id);
+                                return Ok(());
+                            }
+                            None => {
+                                error!(meta.time(); "received response while no query is sent");
+                                return Ok(());
+                            }
+                        };
+                        // unset pending
+                        dispatcher.push(P2pNetworkRpcAction::PrunePending {
+                            peer_id: *peer_id,
+                            stream_id: *stream_id,
+                        });
+
+                        if let Err(e) =
+                            rpc_response_effects(*peer_id, &query_header, bytes, dispatcher)
```

Comment:
Just in case, the logic inside this functions does not depend on the `PrunePending` above having been processed already right?

Reply:
nevermind, it doesn't receive as input the state object, so there is no dependency

Reply:
It doesn't depend on `PrunePending`, but  `PrunePending` action is used twice only to set `pending` to `None` this can be move to reducers.